hedera-node/src/main/java/com/hedera/services/store/contracts/HederaWorldState.java

```diff
@@ -407,27 +333,58 @@ public void revert() {
 		@Override
 		public void commit() {
 			final HederaWorldState wrapped = wrappedWorldView();
-			wrapped.provisionalAccountDeletes.addAll(getDeletedAccountAddresses());
-			wrapped.sponsorMap.putAll(sponsorMap);
+			var repository = wrapped.repositoryRoot;
+
+			getDeletedAccountAddresses().forEach(address -> {
+				// account may have been created and deleted within the same transaction.
+				if (repository.isExist(address.toArray())) {
+					var account = repository.getAccount(address.toArray());
+					// selfdestruct adjusts balances to zero
+					account.setBalance(BigInteger.ZERO);
+					validateTrue(!repository.isDeleted(address.toArray()), FAIL_INVALID);
+					account.setDeleted(true);
+				}
+			});
 
 			for (final UpdateTrackingAccount<WorldStateAccount> updated : getUpdatedAccounts()) {
+
 				final WorldStateAccount origin = updated.getWrappedAccount();
+				byte[] address = updated.getAddress().toArray();
+				long balance = updated.getBalance().toLong();
+				final var accountId = accountParsedFromSolidityAddress(address);
+				AccountState updatedAccount;
+
+				if (repository.isExist(address)) {
+					updatedAccount = repository.getAccount(address);
+					updatedAccount.setBalance(BigInteger.valueOf(updated.getBalance().toLong()));
+				} else {
+					repository.delete(updated.getAddress().toArray());
+					updatedAccount = repository.createAccount(address);
+					updatedAccount.setSmartContract(true);
+
+					wrapped.createdContracts.add(EntityIdUtils.asContract(accountId));
+				}
+
+				updatedAccount.setBalance(BigInteger.valueOf(balance));
+
+				if (getDeletedAccountAddresses().contains(address)) {
+					updatedAccount.setDeleted(true);
```

Comment:
Will a given deleted account be present in both `getDeletedAccountAddresses` and `getUpdatedAccounts`? If not, this if is unnecessary.